### PR TITLE
feat: Make integration tests rust unit tests

### LINKED ISSUE #2881: Make integration tests rust unit tests
I have some preliminary work here https://github.com/ry/deno/tree/itest

The idea is to remove all the `//tests/*.test` files and instead define a macro in Rust to declare a unit test for each. It would look something like this:

```rust
itest!(_001_hello {
  args: "run --reload 001_hello.js",
  output: "tests/001_hello.js.out",
});

itest!(_002_hello {
  args: "run --reload 002_hello.ts",
  output: "tests/002_hello.ts.out",
});

itest!(_003_relative_import {
  args: "run --reload 003_relative_import.ts",
  output: "tests/003_relative_import.ts.out",
});
```
See https://github.com/ry/deno/blob/itest/cli/integration_tests.rs for full code

The benefit of this is that they can be run in parallel and we will have less custom test infrastructure to maintain.

If someone is interested in finishing off my branch, let me know. I think it should be close to working - but needs some clean ups.

### PR DESCRIPTION
**Pull request by bartlomieju**
Closes #2881

Todo:
- [x] add support for stdin in test case
- [ ] run `tools/http_server.py` while tests are running (it seems Rust doesn't support running custom scripts before `cargo test`, need to investigate)
- [x] multiplex `stdout`/`stderr` of child process to capture output properly ~~(Rust doesn't support it out-of-the-box 🙄, one solution is to create temporary files, or use lib like [`duct`](https://docs.rs/duct/0.12.0/duct/) that can handle that easily)~~ (used [`os_pipe`](https://docs.rs/os_pipe/0.8.2/os_pipe/) crate for that)

Found problem:
When integration test crashed it hanged the subcommand
```
thread 'tokio-runtime-worker-0' panicked at 'Can't get ErrorKind for Error(Hyper(Error(Connect, Os { code: 61, kind: ConnectionRefused, message: "Connection refused" })), "http://localhost:4545/tests/subdir/mod2.ts")', cli/deno_error.rs:272:9
thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: Canceled', src/libcore/result.rs:999:5
thread 'tokio-runtime-worker-1' panicked at 'unexpected state while aborting task: Complete', /Users/biwanczuk/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-threadpool-0.1.15/src/task/mod.rs:216:21
test integration_tests::error_016_dynamic_import_permissions2 ... test integration_tests::error_016_dynamic_import_permissions2 has been running for over 60 seconds
test worker::tests::execute_006_url_imports ... test worker::tests::execute_006_url_imports has been running for over 60 seconds
```

### CODE REVIEW

**Comment by bartlomieju** [comment]
So this is working now for me locally - I updated a couple of integration tests to make them independent of each other.

This PR adds `os_pipe` dev dependency.  

There is still the issue of running `http_server` that I'm not yet sure how to tackle. When you run `./tools/test.py` it's all good, but if you try to run `./target/debug/cli_test` you need to spawn the server manually. I did some research on running custom scripts/command before `cargo test` and all answers direct to testing frameworks that can handle that, but there's no such functionality built-into `cargo`. Are you ok with this approach?

@ry please review

File: cli/integration_tests.rs [hunk 1]
```diff
@@ -0,0 +1,615 @@
...
+    let (mut reader, writer) = pipe().unwrap();
+    let mut command = Command::new(bin);
+    command.args(args);
+    command.current_dir(root.join("tests"));
+    command.stdin(Stdio::piped());
+    command.stderr(Stdio::null());
+
+    if self.check_stderr {
+      let writer_clone = writer.try_clone().unwrap();
+      command.stderr(writer_clone);
```
Comment (ry) [inline cli/integration_tests.rs]: 👍

File: cli/integration_tests.rs [hunk 2]
```diff
@@ -0,0 +1,615 @@
...
+  // Iterate through the fixture lists, testing each one
+  for (pattern, string, expected) in fixtures {
+    let actual = wildcard_match(pattern, string);
+    dbg!(pattern, string, expected);
+    assert_eq!(actual, expected);
+  }
+
+  // TODO different wild cards?
+  // assert!(pattern_match("foo[BAR]baz", "foobarbaz", "[BAR]"));
+  // assert!(!pattern_match("foo[BAR]baz", "foobazbar", "[BAR]"));
```
Comment (ry) [inline cli/integration_tests.rs]: Remove these? Seems like you have the test below

Comment (bartlomieju) [reply cli/integration_tests.rs]: Yikes, fixed!

**Comment by bartlomieju** [comment]
Rebased on top of `master`

**Comment by bartlomieju** [comment]
Test are failing on Windows, I expect it's because of CLRF... I'll update it tomorrow

**Comment by bartlomieju** [comment]
Thanks @ry, but most of the hard work was done by you :)

**Pull request merged**